src/box/xlog_reader.c

```diff
@@ -0,0 +1,518 @@
+#include <stdint.h>
+#include <small/ibuf.h>
+#include <small/lsregion.h>
+
+#include "xlog_reader.h"
+
+#include "cbus.h"
+#include "fiber.h"
+#include "fiber_cond.h"
+#include "memory.h"
+#include "tt_static.h"
+#include "xlog.h"
+
+/**
+ * Stream message. It is used to pass batches with xlog requests from reader
+ * thread to TX and back to be freed.
+ */
+struct xlog_reader_stream_msg {
+	/** Base class. */
+	struct cmsg base;
+	/** Xlog reader instance. */
+	struct xlog_reader *reader;
+	/**
+	 * Requests of the batch.
+	 * Request is object of `struct recovery_request`.
+	 */
+	struct stailq requests;
+	/** Position of the batch in reader batches ring buffer. */
+	unsigned pos;
+	/** Error reading xlog. If it is set requests can be not empty. */
+	struct error *error;
+	/**
+	 * Set if xlog EOF is reached. The `error` can be set as well
+	 * in this case.
+	 */
+	bool eof;
+};
+
+/**
+ * Control message. It is used to send stop or finished message. See also
+ * xlog_reader::stop_msg and xlog_reader::finished_msg.
+ */
+struct xlog_reader_control_msg {
+	/** Base class. */
+	struct cmsg base;
+
+	/** Xlog reader instance. */
+	struct xlog_reader *reader;
+};
+
+/** Number xlog_reader_stream_msg messages in ring buffer. See xlog_reader. */
+#define MSG_COUNT 2
+
+/**
+ * Xlog reader, uses a designated thread to read xlogs and parse them to
+ * requests (see `struct recovery_request`).
+ *
+ * Snapshot xlogs are written in batches of size about 128k
+ * (`XLOG_TX_AUTOCOMMIT_THRESHOLD`). The reader thread reads and parses
+ * the entire batch and send it to TX. There the requests can be obtained
+ * using `xlog_reader_next`. After the batch is read entirely in TX the
+ * batch is sent back to reader thread to be freed.
+ *
+ * To organize flow control between reader and TX threads the number of batches
+ * passing between threads is limited to `MSG_COUNT` and they are organized
+ * into a ring buffer.
+ */
+struct xlog_reader {
```

Comment:
Too big struct. Let's try to split it. Maybe, we could make some variables local to the thread function?

Also, please consider switching to a different model - when the tx thread sends requests to the reader thread to read batches. Maybe, it'd simplify things (not sure though).